cupy/core/core.pyx

```diff
@@ -114,37 +114,40 @@ cdef class ndarray:
 
     """
 
-    def __init__(self, shape, dtype=float, memptr=None, order='C'):
+    def __init__(self, shape, dtype=float, memptr=None, strides=None,
+                 order='C'):
         cdef Py_ssize_t x, itemsize
         cdef tuple s = internal.get_size(shape)
-        self._shape.reserve(len(s))
-        self.size = 1
+        cdef int order_char = 'C' if order is None else _normalize_order(order)
+
+        # `strides` is prioritized over `order`, but invalid `order` should be
+        # checked even if `strides` is given.
+        if order_char not in ('C', 'F'):
+            raise TypeError('order not understood. order={}'.format(order))
+
+        # Check for erroneous shape
         for x in s:
             if x < 0:
                 raise ValueError('Negative dimensions are not allowed')
-            self._shape.push_back(x)
-            self.size *= x
+
+        # dtype
         self.dtype, itemsize = _dtype.get_dtype_with_itemsize(dtype)
 
+        # Store shape and strides
+        if strides is not None:
+            self._set_shape_and_strides(shape, strides, True, True)
```

Comment:
Ah yes, I've found that problem but just ignored it because it seemed not easy to replicate NumPy behavior. I agree with restricting specificaton.
Should we raise `ValueError`? (another candidate would be `NotImplementedError`?)

Reply:
`ValueError` looks good.

Reply:
Fixed.